File: src/main/java/io/github/mizar107/zapegcitizens/brain/HarvestPolicy.java
package io.github.mizar107.zapegcitizens.brain;

import io.github.mizar107.zapegcitizens.chat.TurkishFold;

import java.util.Locale;
import java.util.regex.Pattern;

/**
 * Vanilla harvest rules the planner cannot treat as owner blockers.
 *
 * <p>Logs and similar hand-breakable blocks can be punched. An axe is optional
 * speed, never a requirement. Stone and ore still need a pickaxe.
 *
 * <p>The keyword nets are bilingual (the live server speaks Turkish): every
 * English family carries its Turkish synonyms ("balta" for axe, "kürek" for
 * shovel, "makas" for shears, "odun/kütük" for wood/logs). Matching runs over
 * {@link TurkishFold#fold} output so {@code İ}/{@code I}/{@code ı} all compare
 * as a plain {@code i}, and the patterns are written in that folded form.
 */
public final class HarvestPolicy {

    static final String HAND_HARVEST_ANSWER =
            "Wood and other hand-breakable blocks can be punched. Do not wait for an "
                    + "axe. Call mine on the log or wood block ids now. Equip an axe only if "
                    + "one is already in inventory.";

    private static final int KEYWORD_FLAGS = Pattern.CASE_INSENSITIVE
            | Pattern.UNICODE_CASE | Pattern.UNICODE_CHARACTER_CLASS;

    private static final Pattern WOOD_GOAL = Pattern.compile(
            "\\b(wood|logs?|trees?|chop|lumber|timber|"
                    + "oak|spruce|birch|jungle|acacia|dark.?oak|mangrove|cherry|"
                    + "crimson|warped|hyphae|stems?|"
                    + "odun\\w*|kütü(?:k|ğ)\\w*|kutuk\\w*|ağa(?:c|ç)\\w*|agac\\w*|"
                    + "kereste\\w*|tahta\\w*)\\b",
            KEYWORD_FLAGS);
    private static final Pattern OPTIONAL_TOOL = Pattern.compile(
            "\\b(axes?|hatchets?|shovels?|shears?|"
                    + "balta\\w*|kürek\\w*|kurek\\w*|makas\\w*|nacak\\w*)\\b",
            KEYWORD_FLAGS);
    private static final Pattern PICKAXE = Pattern.compile(
            "\\b(pickaxes?|kazma\\w*)\\b", KEYWORD_FLAGS);
    private static final Pattern PUNCHABLE_QUESTION = Pattern.compile(
            "\\b(wood|logs?|trees?|leaves?|dirt|sand|gravel|punch|"
                    + "odun\\w*|kütü(?:k|ğ)\\w*|kutuk\\w*|ağa(?:c|ç)\\w*|agac\\w*|"
                    + "yaprak\\w*|toprak\\w*|kum\\w*|çakil\\w*|cakil\\w*|yumruk\\w*)\\b",
            KEYWORD_FLAGS);

    private HarvestPolicy() {}

    public static boolean isOptionalHarvestToolRequest(String goal, String question) {
        String goalText = TurkishFold.fold(goal);
        String questionText = TurkishFold.fold(question);
        if (!OPTIONAL_TOOL.matcher(questionText).find()) {
            return false;
        }
        if (PICKAXE.matcher(questionText).find()
                && !OPTIONAL_TOOL.matcher(PICKAXE.matcher(questionText).replaceAll(" ")).find()) {
            return false;
        }
        return WOOD_GOAL.matcher(goalText).find()
                || PUNCHABLE_QUESTION.matcher(questionText).find();
    }

    public static String handHarvestAnswer() {
        return HAND_HARVEST_ANSWER;
    }

    static boolean describesAxeSupply(String inventoryChange) {
        if (inventoryChange == null || inventoryChange.isBlank()) {
            return false;
        }
        String normalized = inventoryChange.toLowerCase(Locale.ROOT);
        return normalized.contains("_axe") || normalized.contains(":axe");
    }
}

File: brain/citizen_brain/harvest_policy.py
"""Server-side harvest rules the planner cannot override.

Vanilla players punch logs, dirt, sand, and similar blocks by hand. An axe is a
speed upgrade, not a capability gate. Stone and ore still need a pickaxe.

The keyword nets are bilingual: the live server speaks Turkish, so every
English family carries its Turkish synonyms ("balta" for axe, "kürek" for
shovel, "makas" for shears, "çapa" for hoe, "orak" for sickle, "odun/kütük"
for wood/logs, "şeker kamışı/kamış" for sugar cane). Matching runs over
:func:`citizen_brain.textfold.fold` output, so ``İ``/``I``/``ı`` all compare as
a plain ``i`` and the patterns are written in that folded form.
"""

from __future__ import annotations

import re

from .textfold import fold

HAND_HARVEST_FAULT = (
    "wood, sugar cane, crops, and other hand-breakable blocks do not require "
    "a harvest tool; call mine with the exact block ids now. Punching with "
    "empty hands is allowed. Equip an axe only if one is already in inventory; "
    "never call job_needs_input for an axe, hoe, shovel, or shears."
)

_WOOD_GOAL = re.compile(
    r"\b(wood|logs?|trees?|chop|lumber|timber|"
    r"oak|spruce|birch|jungle|acacia|dark.?oak|mangrove|cherry|"
    r"crimson|warped|hyphae|stems?|"
    r"odun\w*|kütü(?:k|ğ)\w*|kutuk\w*|ağa(?:c|ç)\w*|agac\w*|kereste\w*|tahta\w*)\b",
    re.IGNORECASE,
)
_OPTIONAL_TOOL = re.compile(
    r"\b(axes?|hatchets?|shovels?|shears?|hoes?|sickles?|"
    r"balta\w*|kürek\w*|kurek\w*|makas\w*|nacak\w*|"
    r"çapa\w*|capa\w*|orak\w*|t[ıi]rm[ıi]k\w*)\b",
    re.IGNORECASE,
)

# Surface plants: hand-harvest is the ONLY correct move (the 2026-08-21 field
# failure was a citizen pausing a sugar-cane job to ask the owner for a çapa,
# then re-asking after the owner said to use bare hands).
_PLANT_GOAL = re.compile(
    r"\b(sugar\s*cane|canes?|reeds?|bamboo|crops?|wheat|carrots?|potato\w*|"
    r"beetroots?|melons?|pumpkins?|"
    r"[şs]eker\s*kam[ıi][şs]\w*|kam[ıi][şs]\w*|bambu\w*|ekin\w*|"
    r"bu[ğg]day\w*|havu[çc]\w*|patates\w*|pancar\w*|kavun\w*|karpuz\w*|"
    r"bal\s*kaba[ğg]\w*|hasat\w*)\b",
    re.IGNORECASE,
)
_PICKAXE_ONLY = re.compile(r"\b(pickaxes?|kazma\w*)\b", re.IGNORECASE)
_PUNCHABLE_QUESTION = re.compile(
    r"\b(wood|logs?|trees?|leaves?|dirt|sand|gravel|punch|"
    r"sugar\s*cane|canes?|reeds?|bamboo|crops?|wheat|"
    r"odun\w*|kütü(?:k|ğ)\w*|kutuk\w*|ağa(?:c|ç)\w*|agac\w*|yaprak\w*|"
    r"toprak\w*|kum\w*|çakil\w*|cakil\w*|yumruk\w*|"
    r"[şs]eker\s*kam[ıi][şs]\w*|kam[ıi][şs]\w*|bambu\w*|ekin\w*|"
    r"bu[ğg]day\w*|hasat\w*)\b",
    re.IGNORECASE,
)


def optional_harvest_tool_fault(goal: str | None, question: str | None) -> str | None:
    """Return planner-fault text when job_needs_input asks for a punch-optional tool."""
    goal_text = fold(goal)
    question_text = fold(question)
    if not _OPTIONAL_TOOL.search(question_text):
        return None
    if _PICKAXE_ONLY.search(question_text) and not _OPTIONAL_TOOL.search(
        _PICKAXE_ONLY.sub(" ", question_text)
    ):
        return None
    if (
        _WOOD_GOAL.search(goal_text)
        or _PLANT_GOAL.search(goal_text)
        or _PUNCHABLE_QUESTION.search(question_text)
    ):
        return HAND_HARVEST_FAULT
    return None
